### PR TITLE
Support IngressClassName in TransportServer

### PR DESCRIPTION
**Pull request by pleshakov**
### Proposed changes
The TransportServer resource now includes the `spec.ingressClassName`, which users can set to choose which NGINX Ingress Controller they want to handle their TransportServer resources.  This is useful when multiple NGINX Ingress Controllers are running in the cluster (for example, one for prod and one for dev).

### CODE REVIEW

File: cmd/nginx-ingress/main.go [hunk 1]
```diff
@@ -76,12 +76,13 @@ var (
...
-	For Kubernetes < 1.18, the Ingress Controller only processes resources that belong to its class
-	- i.e have the annotation "kubernetes.io/ingress.class" equal to the class.
+	For Kubernetes < 1.18, the Ingress Controller only processes resources that belong to its class - 
+	i.e have the annotation "kubernetes.io/ingress.class" (for Ingress resources)
+	or field "ingressClassName" (for VirtualServer/VirtualServerRoute/TransportServer resources) equal to the class.
 	Additionally, the Ingress Controller processes resources that do not have the class set,
 	which can be disabled by setting the "-use-ingress-class-only" flag
 
-	The Ingress Controller processes all the VirtualServer/VirtualServerRoute resources that do not have the "ingressClassName" field for all versions of kubernetes.`)
+	The Ingress Controller processes all the VirtualServer/VirtualServerRoute/TransportServer resources that do not have the "ingressClassName" field for all versions of kubernetes.`)
```
Comment (soneillf5) [inline cmd/nginx-ingress/main.go:85]: Can you clarify this sorry, it processes the resources that *do not* have ingressClassName ?

File: internal/k8s/configuration_test.go [hunk 2]
```diff
@@ -2171,6 +2171,70 @@ func TestAddInvalidTransportServer(t *testing.T) {
 	}
 }
 
+func TestAddTransportServerWithIncorrectClass(t *testing.T) {
```
Comment (soneillf5) [inline internal/k8s/configuration_test.go:2174]: Great test!

File: internal/k8s/controller_test.go [hunk 3]
```diff
@@ -229,108 +230,94 @@ func TestHasCorrectIngressClass(t *testing.T) {
 	}
 }
 
-func TestHasCorrectIngressClassVS(t *testing.T) {
-	ingressClass := "ing-ctrl"
-	lbcIngOnlyTrue := &LoadBalancerController{
-		ingressClass:        ingressClass,
+func deepCopyWithIngressClass(obj interface{}, class string) interface{} {
```
Comment (soneillf5) [inline internal/k8s/controller_test.go:233]: Not sure about this function, all the paths looks the same.
Can it be replaced on line 315 by 
```
                objCopy := obj.DeepCopy()
 		objCopy.Spec.IngressClass = class
```

**Review by soneillf5 — CHANGES_REQUESTED** [review]
All good except for the deepCopy() in the test.

File: cmd/nginx-ingress/main.go (hunk 1 above)
Comment (pleshakov) [reply cmd/nginx-ingress/main.go:85]: @soneillf5 
 "The Ingress Controller processes all the VirtualServer/VirtualServerRoute/TransportServer resources with an empty "ingressClassName" field for all versions of kubernetes."  ?

File: internal/k8s/controller_test.go (hunk 3 above)
Comment (pleshakov) [reply internal/k8s/controller_test.go:233]: @soneillf5 
we need to reference a concrete type to apply `objCopy.Spec.IngressClass = class`  , that's why we have a a type switch with 3 cases

Comment (pleshakov) [reply internal/k8s/controller_test.go:233]: if deepCopy is confusing, I can create a concrete type instead, but still do that three times:
```
	case *conf_v1.VirtualServer:
		vs := conf_v1.VirtualServer{}
		vs.Spec.IngressClass = class
		return &vs
```

Comment (soneillf5) [reply internal/k8s/controller_test.go:233]: No that's fine as it is, thanks for the explanation.

**Review by soneillf5 — APPROVED** [review]

File: docs-web/installation/running-multiple-ingress-controllers.md [hunk 4]
```diff
@@ -22,7 +22,7 @@ The default Ingress class of NGINX Ingress Controller is `nginx`, which means th
 **Notes**: 
 * For Kubernetes < 1.18, if the class is not set in an Ingress configuration resource, the Ingress Controller will handle the resource. This is controlled via the `-use-ingress-class-only` argument. 
 * For Kubernetes >= 1.18, if the class is not set in an Ingress resource, Kubernetes will set it to the class of the default Ingress Controller. To make the Ingress Controller the default one, the `ingressclass.kubernetes.io/is-default-class` must be set on the IngressClass resource. See Step 3 *Create an IngressClass resource* of the [Create Common Resources](/nginx-ingress-controller/installation/installation-with-manifests/#create-common-resources) section.
-* For VirtualServer and VirtualServerRoute resources, the Ingress Controller will always handle resources with an empty class.
+* For VirtualServer, VirtualServerRoute, and TransportServer resources the Ingress Controller will always handle resources with an empty class.
```
Comment (lucacome) [inline docs-web/installation/running-multiple-ingress-controllers.md:25]:
```suggestion
* For VirtualServer, VirtualServerRoute and TransportServer resources the Ingress Controller will always handle resources with an empty class.
```

**Review by lucacome — APPROVED** [review]

**Comment by pstackle** [comment]
@pleshakov Is there an estimate of when this change will be included in a release?

**Comment by pleshakov** [comment]
Hi @pstackle 

it will go into 1.11.0, scheduled for the end of march

**Pull request merged**